### PR TITLE
Fix double iteration bug when resumed from a checkpoint.

### LINKED ISSUE #19427: calling iter twice messes up dataloaders with queues
### Bug description

This bug has reappeared https://github.com/Lightning-AI/pytorch-lightning/issues/18414 

We now call iter() twice in different places:
- https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/loops/fit_loop.py#L263
- https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/loops/training_epoch_loop.py#L171C1-L172C1

### What version are you seeing the problem on?

v2.1

### How to reproduce the bug

```python
import multiprocessing as mp
from queue import Queue
from typing import Iterator

import numpy as np
from lightning import Trainer
from lightning.pytorch.demos.boring_classes import BoringModel
from torch.utils.data import DataLoader, IterableDataset


class QueueDataset(IterableDataset):
    def __init__(self, queue: Queue) -> None:
        super().__init__()
        self.queue = queue

    def __iter__(self) -> Iterator:
        for k in range(5):
            print(f"getting {k}")
            tensor, index = self.queue.get(timeout=10)
            print(f"got {index}")
            yield tensor


if __name__ == "__main__":
    q = mp.Queue()
    arr = np.random.random([1, 32]).astype(np.float32)
    for ind in range(10):
        q.put((arr, ind))
    max_epoch = 1
    dataloader = DataLoader(QueueDataset(q), num_workers=1, batch_size=None, persistent_workers=True)
    trainer = Trainer(max_epochs=max_epoch, enable_progress_bar=False, devices=1)
    trainer.fit(BoringModel(), dataloader)
    trainer.save_checkpoint("model.ckpt")

    # q now has the next 5 elems in
    # resuming training we will hit the double iter() issue
    dataloader = DataLoader(QueueDataset(q), num_workers=1, batch_size=None, persistent_workers=True)
    trainer = Trainer(max_epochs=max_epoch + 1, enable_progress_bar=False, devices=1)
    trainer.fit(BoringModel(), dataloader, ckpt_path="model.ckpt")
```

### Error messages and logs

relevant logs are:
```
# first epoch all good
getting 0
got 0
getting 1
got 1
getting 2
got 2
getting 3
got 3
getting 4
got 4

# second epoch we start getting from the queue twice!
# from fit loop iter()
getting 0
got 5
getting 1
got 6
getting 2
got 7
# from training_epoch loop iter()
getting 0
got 8
getting 1
got 9
getting 2
```

### Environment

lighting==2.1.4

### More info

_No response_

cc @justusschock @awaelchli @carmocca

### PR DESCRIPTION
**Pull request by sudiptob2**
## What does this PR do?

This PR fixes the double `iter()` bug discussed in #19427

Fixes #19427

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--20775.org.readthedocs.build/en/20775/

### CODE REVIEW

**Review by Borda — APPROVED** [review]

File: tests/tests_pytorch/loops/test_double_iter_in_iterable_dataset.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import multiprocessing as mp
+import os
+from collections.abc import Iterator
+from pathlib import Path
+from queue import Queue
+
+import numpy as np
+from torch.utils.data import DataLoader, IterableDataset
+
+from lightning import Trainer
```
Comment (Borda) [inline tests/tests_pytorch/loops/test_double_iter_in_iterable_dataset.py:27]:
```suggestion
from lightning.pytorch import Trainer
```

**Comment by sudiptob2** [comment]
@Borda Thanks for the review! Let me know if there's anything you'd like me to change. Otherwise, can we go ahead and merge this?

File: tests/tests_pytorch/loops/test_double_iter_in_iterable_dataset.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    return trainer
+
+
+def test_resume_training_with(tmp_path):
+    """Test resuming training from checkpoint file using a IterableDataset."""
+    queue = create_queue()
+    max_epoch = 2
+    ckpt_path = tmp_path / "model.ckpt"
+    trainer = train_model(queue, max_epoch, ckpt_path)
+    assert trainer is not None
```
Comment (Borda) [inline tests/tests_pytorch/loops/test_double_iter_in_iterable_dataset.py:73]: how or when it would return trainer as None?

Comment (sudiptob2) [reply tests/tests_pytorch/loops/test_double_iter_in_iterable_dataset.py:73]: Actually followed the pattern mentioned in this PR https://github.com/Lightning-AI/pytorch-lightning/pull/20705. I can remove this check.

Comment (sudiptob2) [reply tests/tests_pytorch/loops/test_double_iter_in_iterable_dataset.py]: updated the test

**Comment by sudiptob2** [comment]
Hey, @Borda just wondering if anything is blocking this PR from merging?

File: src/lightning/pytorch/loops/fit_loop.py [hunk 3]
```diff
@@ -204,6 +204,11 @@ def skip(self) -> bool:
         # so we cannot use it solely
         return self.done or self.trainer.limit_train_batches == 0
 
+    @property
+    def _is_resuming(self) -> bool:
+        """Whether we're resuming training from a checkpoint."""
+        return self._loaded_from_state_dict
```
Comment (Borda) [inline src/lightning/pytorch/loops/fit_loop.py]: Looking at how the `_loaded_from_state_dict` is used, and there is no direct wrap for loading, meaning set it as true/false when loading starts, and switch back when it ends so for that we my need to add another/new attribute

Comment (sudiptob2) [reply src/lightning/pytorch/loops/fit_loop.py]: Hi! Thanks for taking a look at it. I am not sure why we need that. `_loaded_from_state_dict` is already made True and False when necessary so I am using `_loaded_from_state_dict` to detect checkpoint resumption.
I might have missed something, could you please suggest what changes we need here?

https://github.com/Lightning-AI/pytorch-lightning/blob/1e88899a51c78534aa6897489829be94d76ef28e/src/lightning/pytorch/loops/loop.py#L89

https://github.com/Lightning-AI/pytorch-lightning/blob/1e88899a51c78534aa6897489829be94d76ef28e/src/lightning/pytorch/loops/loop.py#L102-L105

Comment (Borda) [reply src/lightning/pytorch/loops/fit_loop.py]: I see `_loaded_from_state_dict`, but my point is if needed, let's add new variable starting with _ whhci would reliably describe this status instead of trying recycle already existing one whhci does not have that good a fit

Comment (sudiptob2) [reply src/lightning/pytorch/loops/fit_loop.py]: make sense, will update shortly

Comment (sudiptob2) [reply src/lightning/pytorch/loops/fit_loop.py]: @Borda just update the PR, please take a look

**Comment by sudiptob2** [comment]
@Borda, I noticed a few mypy checks are failing, but they don't seem related to the changes in this PR. Should we ignore them for now? Let me know the next steps.

File: src/lightning/pytorch/loops/loop.py [hunk 4]
```diff
@@ -38,6 +39,11 @@ def restarting(self, restarting: bool) -> None:
             if isinstance(loop, _Loop):
                 loop.restarting = restarting
 
+    @property
+    def is_resuming(self) -> bool:
+        """Whether we're resuming training from a checkpoint."""
+        return self._resuming_from_checkpoint
+
```
Comment (bhimrazy) [inline src/lightning/pytorch/loops/loop.py]: This variable `_resuming_from_checkpoint` seems a bit redundant. From the naming and context, it looks like `_loaded_from_state_dict` might already be serving a similar purpose.  wdyt @sudiptob2  ?

cc: @Borda

Comment (sudiptob2) [reply src/lightning/pytorch/loops/loop.py]: Hey @bhimrazy, thanks for taking a look at the PR. In my initial implementation, I had reused `_loaded_from_state_dict`. However, as @Borda suggested,  a new variable would be useful to describe the status of this **resuming** case reliably.
Please see the review comments in this thread: https://github.com/Lightning-AI/pytorch-lightning/pull/20775#discussion_r2131892490

Comment (bhimrazy) [reply src/lightning/pytorch/loops/loop.py]: Oh, okay. Thanks for letting me know.

Comment (sudiptob2) [reply src/lightning/pytorch/loops/loop.py]: Just wanted to ask, do you have any updates on when we can expect this fix to be merged? This is quite important issue for several users.

Comment (bhimrazy) [reply src/lightning/pytorch/loops/loop.py]: Actually, I’m not sure either. But I can help bring this PR to the forefront to get some attention on it, incase missed.

Comment (sudiptob2) [reply src/lightning/pytorch/loops/loop.py]: Great! That would be very helpful.

**Review by bhimrazy — APPROVED** [review]

**Comment by sudiptob2** [comment]
@bhimrazy, thanks for approving the PR. I can see that a few checks are timed out. Any idea why?

**Comment by sudiptob2** [comment]
Just checking in, any hope of moving this forward?

**Comment by deependujha** [comment]
Hi @sudiptob2

Sorry for the delay and thanks for your patience.
I’m checking out the PR now and will update you shortly.

File: .github/workflows/ci-tests-pytorch.yml [hunk 5]
```diff
@@ -75,7 +75,7 @@ jobs:
           - { os: "macOS-14", pkg-name: "pytorch", python-version: "3.12", pytorch-version: "2.7" }
           - { os: "ubuntu-22.04", pkg-name: "pytorch", python-version: "3.12", pytorch-version: "2.7" }
           - { os: "windows-2022", pkg-name: "pytorch", python-version: "3.12", pytorch-version: "2.7" }
-    timeout-minutes: 50
+    timeout-minutes: 60
```
Comment (deependujha) [inline .github/workflows/ci-tests-pytorch.yml]:
```suggestion
    timeout-minutes: 50
```

**Comment by deependujha** [comment]
Hi @sudiptob2,

I’m seeing an odd issue: all tests pass on macOS 14 with Python 3.12 and 2.7 in under 30 minutes, but the job still hangs until the 50‑minute timeout and then fails.

I suspect the issue to be in the test file,  possibly related to`queue` handling or `multiprocessing` (spawn/fork).

**Comment by deependujha** [comment]
Thanks, @sudiptob2, for your contribution and for making it even better!

**Comment by sudiptob2** [comment]
@deependujha  Thanks to you for the hard work with testing and CI fixes. Really appreciate 🚀
